pkg/controller/direct/dataproc/job_controller.go

```diff
@@ -0,0 +1,473 @@
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +tool:controller
+// proto.service: google.cloud.dataproc.v1.JobController
+// proto.message: google.cloud.dataproc.v1.Job
+// crd.type: DataprocJob
+// crd.version: v1alpha1
+
+package dataproc
+
+import (
+	"context"
+	"fmt"
+
+	dataproc "cloud.google.com/go/dataproc/v2/apiv1"
+	pb "cloud.google.com/go/dataproc/v2/apiv1/dataprocpb"
+	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	krm "github.com/GoogleCloudPlatform/k8s-config-connector/apis/dataproc/v1alpha1"
+	refs "github.com/GoogleCloudPlatform/k8s-config-connector/apis/refs/v1beta1"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/config"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/directbase"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/controller/direct/registry"
+)
+
+const (
+	// Dataproc jobs cannot be updated.
+	// Ref: https://cloud.google.com/dataproc/docs/reference/rest/v1/projects.regions.jobs/patch
+	// "Only the job state and labels are mutable."
+	// State is managed by the service, and labels are metadata.
+	// We consider the core job configuration immutable.
+	dataprocJobErrCannotUpdate = "dataproc jobs cannot be updated"
+)
+
+func init() {
+	registry.RegisterModel(krm.DataprocJobGVK, NewDataprocJobModel)
+}
+
+func NewDataprocJobModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &dataprocJobModel{config: *config}, nil
+}
+
+var _ directbase.Model = &dataprocJobModel{}
+
+type dataprocJobModel struct {
+	config config.ControllerConfig
+}
+
+func (m *dataprocJobModel) client(ctx context.Context) (*dataproc.JobControllerClient, error) {
+	opts, err := m.config.RESTClientOptions()
+	if err != nil {
+		return nil, err
+	}
+
+	return dataproc.NewJobControllerRESTClient(ctx, opts...)
+}
+
+func (m *dataprocJobModel) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.DataprocJob{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	projectRef, err := refs.ResolveProject(ctx, reader, obj.GetNamespace(), obj.Spec.ProjectRef)
```

Comment:
Not sure why we need to resolve the project and location here, as they should already be included in `NewJobIdentity`.